main.go

```diff
@@ -181,10 +182,96 @@ func (m *memo) list(n *node) {
 	}
 }
 
+type dfsStatus int
+
+const (
+	white dfsStatus = iota
+	gray
+	black
+)
+
+type dfsInfo struct {
+	Me     *class    // this class
+	Parent *class    // parent
+	D      int       // discovery time
+	F      int       // finished visiting time
+	Color  dfsStatus // WHITE (not discovered), GRAY (not visited), BLACK (done)
+}
+
+type dfsInfoList []*dfsInfo
+
+func (m *memo) DFS() dfsInfoList {
+	state := make(map[*class]*dfsInfo, len(m.classes))
+	for _, c := range m.classes {
+		state[c] = &dfsInfo{c, nil, -1, -1, white}
+	}
+	t := 0
+
+	res := make([]*dfsInfo, 0, len(m.classes))
+	for _, c := range m.classes {
+		if state[c].Color == white {
+			m.dfsVisit(c, &t, state)
+		}
+		res = append(res, state[c])
+	}
+	return res
+}
+
+func (m *memo) dfsVisit(c *class, t *int, state map[*class]*dfsInfo) {
+	*t++
+	state[c].D = *t
+	state[c].Color = gray
+
+	for _, n := range c.nodes {
+		for i := 0; i < 2; i++ {
+			var v *node
+			if i == 0 {
+				v = n.left
+			} else {
+				v = n.right
+			}
+			if v == nil {
+				continue
+			}
+
+			vc := m.classes[m.classMap[v.class]]
+			if state[vc].Color == white {
+				state[vc].Parent = c
+				m.dfsVisit(vc, t, state)
+			}
+		}
+	}
+
+	state[c].Color = black
+	*t++
+	state[c].F = *t
+}
+
+type byFDesc []*dfsInfo
+
+func (a byFDesc) Len() int           { return len(a) }
+func (a byFDesc) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byFDesc) Less(i, j int) bool { return a[i].F >= a[j].F }
+
+func (m *memo) topoSort() []*class {
+
+	dfs := m.DFS()
+	sort.Sort(byFDesc(dfs))
```

Comment:
Agreed. It cuts a few lines on an area that's not really our focus now. I'll change.